// containers/register/index.js
import React, {Component} from 'react'
import {connect} from 'react-redux'
import classNames from 'classnames'
import {Link} from 'react-router'

import {
    fetchClients
} from 'actions'
import {getClients} from 'selectors'

class Register extends Component {
    constructor (props) {
        super (props);

        this.state = {
            popupOpen: false
        };
    }

    componentDidMount () {
        this.props.fetchClients()
    }

    togglePopup (status) {
        this.setState({
            popupOpen: status
        });
    }

    renderClient (client, index) {
        return (
            <Link key={index} to={"/client/" + `${client.id}`} className="register-list__item">
                <div className="register-list__name">{client.name}</div>
                <div className="register-list__surname">{client.secondName}</div>
            </Link>
        )
    }

    render () {
        const {clients} = this.props;
        const formClass = classNames({
            'form': true,
            'form--active': this.state.popupOpen
        });

        return (
            <div className="register">
                <div
                    onClick={this.togglePopup.bind(this, true)}
                    className="register__btn">
                    Регистрация клиента
                </div>
                <div className="register-list">
                    {clients.map((client, index) => this.renderClient(client, index))}
                </div>
                <form className={formClass}>
                    <div className="form__header">
                        <h2 className="form__title">Регистрация клиента</h2>
                        <div
                            onClick={this.togglePopup.bind(this, false)}
                            className="form__close">
                        </div>
                    </div>
                    <div className="form__main">
                        <div className="form__box form__box--left">
                            <div className="form__row">
                                <label htmlFor="" className="form__label-text">Имя</label>
                                <input type="text" className="form__input-text"/>
                            </div>
                            <div className="form__row">
                                <label htmlFor="" className="form__label-text">Фамилия</label>
                                <input type="text" className="form__input-text"/>
                            </div>
                            <div className="form__row">
                                <label htmlFor="" className="form__label-text">Email</label>
                                <input type="text" className="form__input-text"/>
                            </div>
                            <div className="form__row">
                                <label htmlFor="" className="form__label-text">Телефон</label>
                                <input type="text" className="form__input-text"/>
                            </div>
                        </div>
                        <div className="form__box form__box--right">
                            <div className="form__row">
                                <label htmlFor="" className="form__label-text">Уезд</label>
                                <input type="text" className="form__input-text"/>
                            </div>
                            <div className="form__row">
                                <label htmlFor="" className="form__label-text">Город / Область</label>
                                <input type="text" className="form__input-text"/>
                            </div>
                            <div className="form__row">
                                <label htmlFor="" className="form__label-text">Адрес</label>
                                <input type="text" className="form__input-text"/>
                            </div>
                            <div className="form__row">
                                <label htmlFor="" className="form__label-text">Индекс</label>
                                <input type="text" className="form__input-text"/>
                            </div>
                        </div>
                    </div>
                    <div className="form__checkblock">
                        <div className="form__checkrow">
                            <input type="checkbox" id="dispatch-email" className="form__input-checkbox"/>
                            <label htmlFor="dispatch-email" className="form__checkbox-ico"></label>
                            <label htmlFor="dispatch-email" className="form__checkbox-label">Рекламная рассылка по э-почте</label>
                        </div>
                        <div className="form__checkrow">
                            <input type="checkbox" id="dispatch-sms" className="form__input-checkbox"/>
                            <label htmlFor="dispatch-sms" className="form__checkbox-ico"></label>
                            <label htmlFor="dispatch-sms" className="form__checkbox-label">Рекламная рассылка по смс</label>
                        </div>
                    </div>
                    <div className="form__btn-wrap">
                        <button className="form__btn">Зарегистрировать</button>
                    </div>
                </form>
                <div
                    onClick={this.togglePopup.bind(this, false)}
                    className="shadow">
                </div>
            </div>
        )
    }
}

const mapStateToProps = (state, ownProps) => ({
    clients: getClients(state, ownProps)
})

const mapDispatchToProps = {
    fetchClients
}

export default connect(mapStateToProps, mapDispatchToProps)(Register)
