// containers/client/index.js
import React, {Component} from 'react'
import {connect} from 'react-redux'
import {Link} from 'react-router'

import {
    getClientById
} from 'selectors'

import {
    updateClient
} from 'actions'

import UpdateForm from 'components/updateForm'

class Client extends Component {
    constructor (props) {
        super (props);

        this.state = {
            update: false
        }
    }

    toggleUpdate () {
        this.setState({
            update: !this.state.update
        })
    }

    renderMain () {
        const { client } = this.props;

        return (
            <div className="personal__main personal__main--start">
                <div className="personal__right">
                    <div className="form__row">
                        <span className="form__label-text">Имя</span>
                        <span className="personal__row-value">{client.name}</span>
                    </div>
                    <div className="form__row">
                        <span className="form__label-text">Фамилия</span>
                        <span className="personal__row-value">{client.secondName}</span>
                    </div>
                    <div className="form__row">
                        <span className="form__label-text">Email</span>
                        <span className="personal__row-value">{client.email}</span>
                    </div>
                    <div className="form__row">
                        <span className="form__label-text">Телефон</span>
                        <span className="personal__row-value">{client.phone}</span>
                    </div>
                    <div className="form__row">
                        <span className="form__label-text">Уезд</span>
                        <span className="personal__row-value">{client.region}</span>
                    </div>
                    <div className="form__row">
                        <span className="form__label-text">Город / Область</span>
                        <span className="personal__row-value">{client.city}</span>
                    </div>
                    <div className="form__row">
                        <span className="form__label-text">Адресс</span>
                        <span className="personal__row-value">{client.address}</span>
                    </div>
                    <div className="form__row">
                        <span className="form__label-text">Индекс</span>
                        <span className="personal__row-value">{client.index}</span>
                    </div>
                </div>
                <div className="personal__right">
                    <div className="form__row">
                        <span className="form__label-text">Статус клиента</span>
                        <span className="personal__row-value">Активный</span>
                    </div>
                    <div className="form__row">
                        <span className="form__label-text">Источник</span>
                        <span className="personal__row-value">Через сайт</span>
                    </div>
                    <div className="form__row">
                        <span className="form__label-text">IP регистрации</span>
                        <span className="personal__row-value"><span className="personal__value--underline">{client.ipInfo.ip}</span> <span>( {client.ipInfo.country} )</span></span>
                    </div>
                    <div className="form__row">
                        <span className="form__label-text">Язык обращения</span>
                        <span className="personal__row-value">{client.language}</span>
                    </div>
                    <div className="form__row">
                        <span className="form__label-text">Пригласил друзей</span>
                        <span className="personal__row-value">0</span>
                    </div>
                    <div className="form__row">
                        <span className="form__label-text">Рекламная рассылка</span>
                        <span className="personal__row-value">
                            <div className="personal__row-value--checkbox">
                                <input type="checkbox" checked={client.dispathEmail} disabled={true}/>
                                <label>э-почта</label>
                            </div>
                            <div className="personal__row-value--checkbox">
                                <input type="checkbox" checked={client.dispathSms} disabled={true}/>
                                <label>смс</label>
                            </div>
                        </span>
                    </div>
                </div>
            </div>
        )
    }

    submitClient (client) {
        if (client) {
            this.props.updateClient(client);
            this.setState({
                update: false
            });
        }
    }

    render () {
        const { client } = this.props;

        return (
            <div className="client">
                <Link to="/" className="back-to-main">
                    Главная
                </Link>
                <div className="personal">
                    <div className="personal__header">
                        <h2 className="personal__title">Данные клиента</h2>
                        <span className="personal_edit" onClick={this.toggleUpdate.bind(this)}>
                            {!this.state.update ? 'Редактировать' : 'Отмена'}
                        </span>
                    </div>
                    {client && !this.state.update && this.renderMain()}
                    {client && this.state.update && <UpdateForm onSubmit={this.submitClient.bind(this)} client={client}></UpdateForm>}
                </div>
                <div className="comments"></div>
            </div>
        )
    }
}

const mapStateToProps = (state, ownProps) => ({
    client: getClientById(state, ownProps.params.id)
})

const mapDispatchToProps = {
    updateClient
}

export default connect(mapStateToProps, mapDispatchToProps)(Client);
